Optimized text for full unicode and some escape sequences

jordan-powers opened the pull request:
Follow-up to #126492 to apply the json parsing optimization to strings containing unicode characters and some backslash-escaped characters.

Supporting backslash-escaped strings is tricky as it requires modifying the string. There are two types of modification: some just remove the backslash (e.g. `\"`, `\\`), and some replace the whole escape sequence with a new character (e.g. `\n`, `\r`, `\u00e5`). In this implementation, the optimization only support the first case--removing the backslash. This is done by making a copy of the data, skipping the backslash. It should still be more optimized than full `String` decoding, but it won't be as fast as non-backslashed strings where we can directly reference the input bytes.


Relates to #129072.

jordan-powers committed:
Add support for some escape sequences in optimizedText

jordan-powers committed:
Extend json parser randomized testing to include escape sequences

jordan-powers committed:
Add full unicode support to optimizedText

jordan-powers committed:
Use single arraylist instance to track backslashes

elasticsearchmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

jordan-powers committed:
Merge remote-tracking branch 'upstream/main' into optimized-text-full-unicode

martijnvg reviewed (approved):
LGTM

I think a review from @elastic/es-core-infra is also required here.

jordan-powers commented on libs/x-content/impl/src/main/java/org/elasticsearch/xcontent/provider/json/ESUTF8StreamJsonParser.java:
```diff
@@ -21,9 +21,14 @@
     protected int stringEnd = -1;
+    protected int stringLength;
+
+    private final List<Integer> backslashes = new ArrayList<>();
```
Unfortunately, Lucene isn't available in `libs/x-content/impl`

alexey-ivanov-es reviewed (approved):
LGTM

jordan-powers committed:
Give up if input buffer is fully consumed

If we completely consume the input buffer before finding a quote
character ending the string, we should return null and fall back to
jackson's native getValueAsString() which has logic to load more data
into the input buffer.

jordan-powers committed:
Merge remote-tracking branch 'upstream/main' into optimized-text-full-unicode
